File: educational-core/src/com/jetbrains/edu/learning/framework/FrameworkLessonManager.java
package com.jetbrains.edu.learning.framework;

import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.jetbrains.edu.learning.courseFormat.FrameworkLesson;
import com.jetbrains.edu.learning.courseFormat.tasks.Task;
import org.jetbrains.annotations.NotNull;

import java.util.Map;

public interface FrameworkLessonManager {
  void prepareNextTask(@NotNull FrameworkLesson lesson, @NotNull VirtualFile taskDir);
  void preparePrevTask(@NotNull FrameworkLesson lesson, @NotNull VirtualFile taskDir);

  void saveSolution(@NotNull Task task, @NotNull Map<String, String> solutions);

  @NotNull
  static FrameworkLessonManager getInstance(@NotNull Project project) {
    return ServiceManager.getService(project, FrameworkLessonManager.class);
  }
}

File: educational-core/src/com/jetbrains/edu/learning/framework/FrameworkLessonManager.kt
package com.jetbrains.edu.learning.framework;

import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.jetbrains.edu.learning.courseFormat.FrameworkLesson;
import com.jetbrains.edu.learning.courseFormat.tasks.Task;
import org.jetbrains.annotations.NotNull;

import java.util.Map;

public interface FrameworkLessonManager {
  void prepareNextTask(@NotNull FrameworkLesson lesson, @NotNull VirtualFile taskDir, boolean showDialogIfConflict);
  void preparePrevTask(@NotNull FrameworkLesson lesson, @NotNull VirtualFile taskDir, boolean showDialogIfConflict);

  void saveExternalChanges(@NotNull Task task, @NotNull Map<String, String> externalState);
  void updateUserChanges(@NotNull Task task, @NotNull Map<String, String> newInitialState);

  @NotNull
  static FrameworkLessonManager getInstance(@NotNull Project project) {
    return ServiceManager.getService(project, FrameworkLessonManager.class);
  }
}
